src/renderer/atlas/AtlasEngine.cpp

```diff
@@ -0,0 +1,1449 @@
+                    // None of these variables need to be initialized.
+                    // features/featureRangeLengths are marked _In_reads_opt_(featureRanges).
+                    // featureRanges is only > 0 when we also initialize all these variables.
+                    DWRITE_TYPOGRAPHIC_FEATURES feature;
+                    const DWRITE_TYPOGRAPHIC_FEATURES* features;
+                    u32 featureRangeLengths;
+#pragma warning(pop)
+                    u32 featureRanges = 0;
+
+                    if (!_api.fontFeatures.empty())
+                    {
+                        feature.features = _api.fontFeatures.data();
+                        feature.featureCount = gsl::narrow_cast<u32>(_api.fontFeatures.size());
+                        features = &feature;
+                        featureRangeLengths = a.textLength;
+                        featureRanges = 1;
+                    }
+
+                    if (_api.clusterMap.size() < a.textLength)
+                    {
+                        _api.clusterMap = Buffer<u16>{ a.textLength };
+                        _api.textProps = Buffer<DWRITE_SHAPING_TEXT_PROPERTIES>{ a.textLength };
+                    }
+
+                    for (auto retry = 0;;)
+                    {
+                        const auto hr = _sr.textAnalyzer->GetGlyphs(
+                            /* textString          */ _api.bufferLine.data() + a.textPosition,
+                            /* textLength          */ a.textLength,
+                            /* fontFace            */ mappedFontFace.get(),
+                            /* isSideways          */ false,
+                            /* isRightToLeft       */ a.bidiLevel & 1,
+                            /* scriptAnalysis      */ &scriptAnalysis,
+                            /* localeName          */ nullptr,
+                            /* numberSubstitution  */ nullptr,
+                            /* features            */ &features,
+                            /* featureRangeLengths */ &featureRangeLengths,
+                            /* featureRanges       */ featureRanges,
+                            /* maxGlyphCount       */ gsl::narrow_cast<u32>(_api.glyphProps.size()),
+                            /* clusterMap          */ _api.clusterMap.data(),
+                            /* textProps           */ _api.textProps.data(),
+                            /* glyphIndices        */ _api.glyphIndices.data(),
+                            /* glyphProps          */ _api.glyphProps.data(),
+                            /* actualGlyphCount    */ &actualGlyphCount);
+
+                        if (hr == HRESULT_FROM_WIN32(ERROR_INSUFFICIENT_BUFFER) && ++retry < 8)
+                        {
+                            // Grow factor 1.5x.
+                            auto size = _api.glyphProps.size();
+                            size = size + (size >> 1);

```

Comment:
This is how you know Leonard is hardcore. *1.5 is an add and another add of a shift right.

Reply:
Honestly surprised you didn't do += size >> 1 :P

Reply:
The first time I wrote this growth code it took me a few days to figure out a good growth strategy. 😄
But that's been long ago and now I just copy-paste these lines into my code whenever I need something like that.

The reason it's 1.5x is because it's less than the golden ratio (1.618x) which means that at some point the sum of all previous allocations exceed the size of the next largest allocation. The idea here is that allocators often use arenas or slabs and if you have a 1.5x growth factor and a starting allocation size of 128 bytes it looks like this:
![image](https://user-images.githubusercontent.com/2256941/141398631-a6aec1a2-7006-4c62-bc02-774413c0d456.png)